segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/DefragmentOperation.java

```diff
@@ -118,14 +119,24 @@
     private final AtomicInteger readAtOffset = new AtomicInteger();
     private final AtomicInteger bytesToRead = new AtomicInteger();
     private final AtomicInteger currentArgIndex = new AtomicInteger();
+    private final AtomicLong currentIndexOffset = new AtomicLong();
 
-    DefragmentOperation(ChunkedSegmentStorage chunkedSegmentStorage, MetadataTransaction txn, SegmentMetadata segmentMetadata, String startChunkName, String lastChunkName, ArrayList<String> chunksToDelete) {
+    DefragmentOperation(ChunkedSegmentStorage chunkedSegmentStorage,
+                        MetadataTransaction txn,
+                        SegmentMetadata segmentMetadata,
+                        String startChunkName,
+                        String lastChunkName,
+                        ArrayList<String> chunksToDelete,
+                        List<ChunkNameOffsetPair> newReadIndexEntries,
```

Comment:
This is a pretty bad pattern. You are given a mutable list (which you don't know who made or who else is using and/or modifying), and you modify it as a result of your execution. This can create innumerable thread safety issues and lead to bugs. 

Please refactor your code to not take in this argument and instead expose it as some getter which returns it after it has executed.

Reply:
fixed.